### ISSUE TITLE
Prometheus alerting reacts to Kubernetes Endpoints changes slowly

### ISSUE THREAD

**Issue by zuzzas** [prometheus/prometheus#7063]
**What did you do?**

Spot instances have shaken up a couple of InstanceGroups of Kubernetes nodes. Alertmanagers got rescheduled. Prometheus continued to send alerts to an old IPs for a couple of hours.

**What did you expect to see?**

Endpoints object got updated, I expected that Prometheus will immediately start sending alerts to a new set of Endpoints.

**What did you see instead? Under which circumstances?**

**Environment**

* System information:

	Linux 4.15.0-1060-aws x86_64

* Prometheus version:

    prometheus, version 2.16.0 (branch: HEAD, revision:     b90be6f32a33c03163d700e1452b54454ddce0ec)
      build user:       root@80fe9dc1d1f6
      build date:       20200303-09:12:22
      go version:       go1.13.8

* Prometheus configuration file:
```yaml
global:
  evaluation_interval: 30s
  scrape_interval: 30s
rule_files:
- /etc/prometheus/rules/prometheus-main-rulefiles-0/*.yaml
scrape_configs:
alerting:
  alert_relabel_configs:
  - source_labels:
    - namespace
    regex: d8-.+|kube-system|kube-nginx-ingress.*|kube-prometheus|kube-openvpn|None
    target_label: tier
    replacement: cluster
  - source_labels:
    - tier
    regex: ^$
    target_label: tier
    replacement: application
  alertmanagers:
  - path_prefix: /
    scheme: http
    kubernetes_sd_configs:
    - role: endpoints
      namespaces:
        names:
        - d8-monitoring
    relabel_configs:
    - action: keep
      source_labels:
      - __meta_kubernetes_service_name
      regex: madison-proxy
    - action: keep
      source_labels:
      - __meta_kubernetes_pod_container_port_number
      regex: "80"
```

* Logs:
```
level=error ts=2020-03-24T18:27:43.070Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=15 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:27:53.072Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=63 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:27:53.072Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=63 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:28:03.072Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=7 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:28:03.072Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=7 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:28:13.073Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=13 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:28:13.073Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=13 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:28:23.074Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=53 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:28:23.074Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=53 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:28:33.075Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=4 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:28:33.075Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=4 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:28:43.075Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=13 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:28:43.075Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=13 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:28:53.077Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=52 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:28:53.077Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=52 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=warn ts=2020-03-24T18:28:57.303Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 152353559 (152354798)"
level=error ts=2020-03-24T18:29:03.077Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=6 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:29:03.077Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=6 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=warn ts=2020-03-24T18:29:05.099Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 152353164 (152354990)"
level=error ts=2020-03-24T18:29:13.078Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=13 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:29:13.078Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=13 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:29:23.079Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=49 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:29:23.079Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=49 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:29:33.080Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=6 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:29:33.080Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=6 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=warn ts=2020-03-24T18:29:35.971Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 152350252 (152355023)"
level=error ts=2020-03-24T18:29:43.080Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=13 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:29:43.080Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=13 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:29:53.083Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=49 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: dial tcp 100.96.80.42:80: i/o timeout"
level=error ts=2020-03-24T18:29:53.083Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=49 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: dial tcp 100.96.68.82:80: i/o timeout"
level=error ts=2020-03-24T18:30:03.084Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=6 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:30:03.085Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=6 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:30:13.085Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=12 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:30:13.085Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=12 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:30:23.087Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.68.82:80/api/v1/alerts count=49 msg="Error sending alert" err="Post http://100.96.68.82:80/api/v1/alerts: context deadline exceeded"
level=error ts=2020-03-24T18:30:23.087Z caller=notifier.go:534 component=notifier alertmanager=http://100.96.80.42:80/api/v1/alerts count=49 msg="Error sending alert" err="Post http://100.96.80.42:80/api/v1/alerts: context deadline exceeded"
```

And here is an `Endpoints` object that was in the following state even earlier than aforeseen logs (`endpoints.kubernetes.io/last-change-trigger-time`):
```
apiVersion: v1
kind: Endpoints
metadata:
  annotations:
    endpoints.kubernetes.io/last-change-trigger-time: "2020-03-24T18:22:11Z"
  creationTimestamp: "2020-02-27T17:17:25Z"
  labels:
    app: madison-proxy
    heritage: deckhouse
    module: prometheus-madison-integration
    prometheus.deckhouse.io/alertmanager: main
  name: madison-proxy
  namespace: d8-monitoring
  resourceVersion: "152354128"
  selfLink: /api/v1/namespaces/d8-monitoring/endpoints/madison-proxy
  uid: 05cd54eb-5985-11ea-bb68-0213e8a285ca
subsets:
- addresses:
  - ip: 100.96.102.43
    nodeName: ip-10-241-119-89.eu-central-1.compute.internal
    targetRef:
      kind: Pod
      name: madison-proxy-2c5bdd9ba9f80394e478714dc299d007182bc49fed6c98sfr
      namespace: d8-monitoring
      resourceVersion: "152354127"
      uid: 14529618-6dfc-11ea-bf38-02b4265706a6
  - ip: 100.96.82.38
    nodeName: ip-10-241-114-228.eu-central-1.compute.internal
    targetRef:
      kind: Pod
      name: madison-proxy-9c6b5f859895442cb645c7f3d1ef647e1ed5388c159atbz2q
      namespace: d8-monitoring
      resourceVersion: "152318686"
      uid: 00700403-6df5-11ea-bf38-02b4265706a6
  - ip: 100.96.96.47
    nodeName: ip-10-241-107-184.eu-central-1.compute.internal
    targetRef:
      kind: Pod
      name: madison-proxy-0450d21f50d1e3f3b3131a07bcbcfe85ec02dd9758b7vx2vm
      namespace: d8-monitoring
      resourceVersion: "152348266"
      uid: edaabb08-6dfa-11ea-bf38-02b4265706a6
  ports:
  - name: http
    port: 80
    protocol: TCP
```

**Comment by roidelapluie** [comment]
Thanks for the report.

cc @brancz for an opinion on this.

**Comment by zuzzas** [comment]
And, yes, I've tried to pin this on the apiserver availability. Ran some tests by suddenly cutting Prometheus from the apiserver via packet drops in a firewall. It always resulted in a ton of error messages on Prometheus' side.

**Comment by brancz** [comment]
I recall us having had a bug around this at some point in the past. Can you validate that the update did indeed reach Prometheus? There `prometheus_sd_kubernetes_events_total` that might help identify that.

**Comment by zuzzas** [comment]
Managed to catch a problem.

Here's the metric. Quite active, as you can see. 
![image](https://user-images.githubusercontent.com/950313/78882503-d65ba200-7a60-11ea-9b7a-eaa26015ff15.png)

Here are two Pods in the Targets view: 
![image](https://user-images.githubusercontent.com/950313/78882603-f68b6100-7a60-11ea-8e67-2985b9ac6fbd.png)

Here's the view of a Namespace's Pods: 
```text
$ kubectl -n d8-cloud-instance-manager get pod -o wide
NAME                                         READY   STATUS    RESTARTS   AGE   IP             NODE                   NOMINATED NODE   READINESS GATES
cluster-autoscaler-6ccf9cd54f-wrxjb          2/2     Running   0          25h   10.244.0.249   p-e2etests-ks-master   <none>           <none>
machine-controller-manager-cc5c96599-lg948   2/2     Running   0          49m   10.244.0.7     p-e2etests-ks-master   <none>           <none>
```

And here is the last clue. Prometheus logs are filled with client-go's reflector's info messages. They are quite frequent, but shouldn't pose a problem (except higher load on apiserver/etcd) since after failed Watch there is a List operation. Perhaps, there is a problem with handling missed Watch delete events?

I've used [this](https://godoc.org/k8s.io/client-go/tools/cache#DeletedFinalStateUnknown) in my controllers to rectify this issue.

```
level=warn ts=2020-04-09T09:31:25.721Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 154739866 (154756428)"
level=warn ts=2020-04-09T09:31:49.726Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 154738657 (154756990)"
level=warn ts=2020-04-09T09:31:58.856Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 154755826 (154756990)"
level=warn ts=2020-04-09T09:32:53.831Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:263: watch of *v1.Pod ended with: too old resource version: 154738917 (154758776)"
level=warn ts=2020-04-09T09:34:10.740Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 154758280 (154761294)"
level=warn ts=2020-04-09T09:35:08.116Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:261: watch of *v1.Endpoints ended with: too old resource version: 154761215 (154764265)"
level=warn ts=2020-04-09T09:36:04.800Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 154757942 (154765096)"
level=warn ts=2020-04-09T09:36:10.819Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 154761783 (154765097)"
level=warn ts=2020-04-09T09:36:44.798Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 154761535 (154765275)"
level=warn ts=2020-04-09T09:36:51.728Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 154763023 (154765275)"
level=warn ts=2020-04-09T09:37:11.966Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 154761783 (154765283)"
level=warn ts=2020-04-09T09:37:56.719Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 154762336 (154765498)"
level=warn ts=2020-04-09T09:38:14.761Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 154761868 (154765531)"
level=warn ts=2020-04-09T09:38:27.843Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:263: watch of *v1.Pod ended with: too old resource version: 154764936 (154765531)"
level=warn ts=2020-04-09T09:39:17.805Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:263: watch of *v1.Pod ended with: too old resource version: 154758362 (154765698)"
level=warn ts=2020-04-09T09:39:45.877Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:281: watch of *v1.Pod ended with: too old resource version: 154763023 (154765698)"
level=warn ts=2020-04-09T09:44:14.171Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/prometheus/prometheus/discovery/kubernetes/kubernetes.go:261: watch of *v1.Endpoints ended with: too old resource version: 154768649 (154770928)"
```

**Comment by roidelapluie** [comment]
cc @brancz can you have a look at this issue again? Thanks

**Comment by hjkatz** [comment]
👋  This exact issue has been happening to us as well so I decided to dig into the whole system (prometheus, config-reloader, client-go, informers, etc...) to try and figure out what's going on, and I have a hypothesis!

Sources:
* Prometheus RoleEndpoints informers: https://github.com/prometheus/prometheus/blob/c9d827c1d019c70adb2e0c6747d1695254ae175d/discovery/kubernetes/kubernetes.go#L378
* SharedInformer docs: https://pkg.go.dev/k8s.io/client-go/tools/cache#SharedInformer
* Google group discussion around shared informers: https://groups.google.com/g/kubernetes-sig-api-machinery/c/PbSCXdLDno0
* Closed k8s PR for updated docs that were never updated: https://github.com/kubernetes/kubernetes/pull/71893
* Referenced re-_list_ period in prom-op code: https://github.com/prometheus-operator/prometheus-operator/blob/a174c8b1281219526db395a669d65a5bac7fe425/pkg/prometheus/operator.go#L269-L278

tl; dr: re-_sync_ and re-_list_ are different things for informers. Resyncs happen automatically for informers, but relists do not. Resyncs will replay any cached events, but not missed events whereas relists will query the api server for an authoritative relist of everything and call `OnUpdate` handler for all listed objects.

A brief search of some related prom code above for the Endpoints informers leads me to believe that no relist events or intervals are being defined, so I think it's possible that prometheus is just missing events and then never relisting to get an authoritative list of all Services and their Endpoints (to then update the `/targets` with the correct Pod IPs).

Overall this problem seems very tricky and I may be headed the wrong direction, nonetheless I felt that putting this information into one place on the internet was a good idea... so here we are 😂 

If somebody more familiar with prometheus' source code and structure could take a look with the idea of re-_list_ being a possible thing that should happen for these informers then that would be amazing. cc @brancz perhaps you could add some clarity to the situation?

**Comment by zuzzas** [comment]
After pondering on this, I think the problem lies in Prometheus losing a stable connection to the apiserver. It's a problem of a particular cluster. You'll see a lot of `watch of *v1.Endpoints ended with: too old resource version: 154761215 (154764265)"`.

@hjkatz, a fine detective work, but the final clue is missing. Re-lists are not defined when working with Informers. They happen automatically when a Watch method (happens after losing streaming connection to apiserver) has set a `resource_version` field that is too old and it's not present in apiserver's watch cache nor in the etcd history.

**Comment by jolusch** [comment]
We are still seeing this issue. Alertmanager pods get rescheduled on a new node and Prometheus keeps sending alerts to the old pod IPs. 
`level=error ts=2021-07-20T04:39:24.484Z caller=notifier.go:527 component=notifier alertmanager=http://100.96.176.58:9093/api/v1/alerts count=4 msg="Error sending alert" err="Post \"http://100.96.176.58:9093/api/v1/alerts\": dial tcp 100.96.176.58:9093: connect: connection refused"`

We are running Prometheus `2.25.0`, Prometheus Operator `0.44.1` and Alertmanager `0.21.0`

**Comment by william-tran** [comment]
I think we've hit this too with v2.26.1, configuring alertmanager with k8s pods service discovery: `level=error ts=*** caller=notifier.go:527 component=notifier alertmanager=http://****:9093/api/v2/alerts count=9 msg="Error sending alert" err="Post \"http://***:9093/api/v2/alerts\": context deadline exceeded"`

**Comment by nielsole** [comment]
I've seen this coincide with a cluster upgrade in GKE. (which will have closed watch connections) The connections from prometheus started timing out presumably on a node pool upgrade(can't tell for certain), so _after_ GKE control plane was upgraded. I suspect there is a bug hidden here beyond the race conditions discussed above.
(v.2.18.1)

**Comment by primeroz** [comment]
For reference i am hitting this as well on prometheus `2.31.1` , updated alertmanager which rescheduled the pods and now prometheus is failing to ship alerts because it is using the old ip. This has been going on for at least **12 hours** since i left it overnight to check if it would fix itself.

**Comment by primeroz** [comment]
Also i did not notice this issue on any other service discovery for my targets, just for the `alertmanager` endpoints in the alertmanager config

**Comment by roidelapluie** [comment]
This is probably a duplicate of https://github.com/prometheus/prometheus/issues/8768

**Comment by roidelapluie** [comment]
I will close this as duplicate of #8768 for which I have produced a fix that should be available in Prometheus 2.37.
